Add more documentation for assessment configuration (#1672)
PrairieLearn/PrairieLearn: docs/assessment/index.md

Reviewer: Should the `uuid` field be linked to our UUID page?

The assessment ID is the full path relative to `assessments`.

## Format specification

| Property                | Type    | Description                                                                                                                                                                    |
| ----------------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `uuid`                  | string  | Unique identifier (UUID). E.g., `"8b4891d6-64d1-4e89-b72d-ad2133f25b2f"`. These can be obtained from [uuidgenerator.net](https://www.uuidgenerator.net) (Required; no default) |

> | `type`                  | string  | Either `"Homework"` or `"Exam"`. ([Details;](#assessment-types) Required; no default)                                                                                          |

Reviewer: Same nit in some other items

> | `type`                  | string  | Either `"Homework"` or `"Exam"`. ([Details;](#assessment-types) Required; no default)                                                                                          |

Reviewer 2: Changed all of these in bbb2e70. There's no longer details links at all.

> | `type`                                                            | string  | Either `"Homework"` or `"Exam"`. ([Details](#assessment-types); Required; no default)                                                     |

Reviewer 2: Nope! Changed in 00f6ebd.

| `title`                 | string  | The title of the assessment (e.g., `"Derivatives and integrals"`). (Required; no default)                                                                                      |
| `set`                   | string  | Which assessment set this belongs to (e.g., `"Quiz"`, `"Practice Quiz"`). (Required; no default)                                                                               |

> | `number`                | string  | The number of the assessment within the set (e.g., `"1"`, `"2B"`). (Required; no default)                                                                                      |

Reviewer: Maybe linking to / talking about the abbreviation would be nice.

> | `number`                | string  | The number of the assessment within the set (e.g., `"1"`, `"2B"`). (Required; no default)                                                                                      |

Reviewer 2: I added a link to that section in bbb2e70.

> | [`module`](#grouping-assessments-by-modules)                      | string  | The module that this assessment belongs to (e.g., `Chapter 3`). (Optional; default: none)                                                 |

Reviewer 3: ```suggestion
| [`module`](#grouping-assessments-by-modules)                      | string  | The module that this assessment belongs to (e.g., `Chapter 3`). (Optional; default: "Default")                                            |
```

Reviewer 2: Changed in a25143d.

| `allowAccess`           | array   | List of access rules. (Optional; default: no student access)                                                                                                                   |
| `zones`                 | array   | Specification of zones and questions. (Optional; default: none)                                                                                                                |
| `text`                  | string  | HTML text shown on the assessment overview page. (Optional; default: none)                                                                                                     |

> | `multipleInstance`      | boolean | Whether to allow students to create whole new attempts at the entire assessment.(Optional; default: `false`)                                                                   |

Reviewer 2: ```suggestion
| `multipleInstance`      | boolean | Whether to allow students to create whole new attempts at the entire assessment. (Optional; default: `false`)                                                                   |
```

| `maxPoints`             | number  | The maximum points that can be earned. (Optional; default: sum of zone max points)                                                                                             |

> | `maxBonusPoints`        | number  | The maximum number of additional points that can be earned beyond `maxPoints` (Optional; default: 0)                                                                           |

Reviewer 2: ```suggestion
| `maxBonusPoints`        | number  | The maximum number of additional points that can be earned beyond `maxPoints`. (Optional; default: 0)                                                                           |
```

> | `maxBonusPoints`        | number  | The maximum number of additional points that can be earned beyond `maxPoints`. (Optional; default: 0)                                                                          |

Reviewer: Do we have text to link to that talks about this?

Reviewer 2: Done in bbb2e70.

> | `shuffleQuestions`                                                | boolean | Whether to randomize the question order (Homework only). (Optional; default: `false`)                                                     |

Reviewer 2: Changed in 3c95960.

> | [`shuffleQuestions`](#question-order-randomization)               | boolean | Whether to randomize the question order (Homework only). (Optional; default: `false`)                                                     |

Reviewer 3: It is no longer Homework-only.

> | [`shuffleQuestions`](#question-order-randomization)               | boolean | Whether to randomize the question order. (Optional; default: `false`)                                                                     |

Reviewer 3: The default actually depends on the assessment type. It's "False" for homework, "True" for exam.

| `groupRoles`            | array   | Array of custom user roles in a group. (Optional; default: none)                                                                                                               |
| `canSubmit`             | array   | A list of group role names that can submit questions in this assessment. Only applicable for group assessments. (Optional; default: none)                                      |
| `canView`               | array   | A list of group role names that can view questions in this assessment. Only applicable for group assessments. (Optional; default: none)                                        |
| `studentGroupCreate`    | boolean | Whether students can create groups. (Optional; default: `false`)                                                                                                               |
| `studentGroupJoin`      | boolean | Whether students can join groups. (Optional; default: `false`)                                                                                                                 |

> | `studentGroupLeave`     | boolean | Whether students can leave groups. (Optional; default: `false`)                                                                                                                |

Author: Done.

See the [reference for `infoAssessment.json`](../schemas/infoAssessment.md) for more information about what can be added to this file.

## Assessment types

Each assessment has a `type`, which must be either `"Homework"` or `"Exam"`.

> Homeworks are designed for formative learning and allow students to infinitely retry different random variants of questions. A persistent student can eventually get 100%.

Reviewer: Can we pull in the mastery diagram here #11839 , or should we do that part of a separate PR?

Reviewer 2: Please open a separate PR if you'd like to see this.

| **Honor pledge**         | Not supported                                                                                       | [Enabled by default](#honor-code)                                                          |
| **Real-time grading**    | Always enabled                                                                                      | [Can be disabled](#disabling-real-time-grading)                                            |

## Assessment naming

> Each assessment has a `title` describing its topic. Additionally, assessments are organized into `sets` (e.g., `Homework`, `Quiz`, `Exam`) and within each set the assessment has a `number`. Depending on the context, assessments are referred to by either an _abbreviation_, a _short name_ or a _long name_. The format of these is:

Reviewer 2: So that things in `code formatting` correspond to the property in JSON, which IMO is nice:

```suggestion
Each assessment has a `title` describing its topic. Additionally, each assessment belongs to a `set` (e.g., `Homework`, `Quiz`, `Exam`) and within each set the assessment has a `number`. Depending on the context, assessments are referred to by either an _abbreviation_, a _short name_ or a _long name_. The format of these is:
```

- Abbreviation = `[Set abbreviation][Number]` (e.g., `Q2` to mean the second quiz).
- Short name = `[Set] [Number]` (e.g., `Quiz 2`).
- Long name = `[Set] [Number]: [Title]` (e.g., `Quiz 2: Coordinates and Vectors`).

> You can select a set name from the list of [standardized assessment sets](../course/index.md#standardized-assessment-sets) or create your [own](../course/index.md#adding-your-own-assessment-sets).

Reviewer 2: Nit, so that the link has a little more text+context for screen readers:

```suggestion
You can select a set name from the list of [standardized assessment sets](../course/index.md#standardized-assessment-sets) or [create your own](../course/index.md#adding-your-own-assessment-sets).
```

The `type` of an assessment does not have to correspond to the `set` it is in, but these are generally compatible. For example, `"type": "Homework"` assessments normally have `"set": "Homework"` or `"set": "Machine Problem"`. On the other hand, `"type": "Exam"` assessments normally have `"set": "Quiz"`, `"set": "Exam"`, or similar.

## Grouping assessments by modules

> Instructors may want to group their assessments by course modules (topics, sections, or chapters in a course). When using `"groupAssessmentsBy" : "Module"` in `infoCourseInstance.json`, instructors can assign an assessment to a specific module by setting the `module` property in `infoAssessment.json`, as illustrated in the example above. If the property `module` is omitted, by default the assessment will have `"module" : "Default"`.

Reviewer: There is some images on the concepts page about how this looks, should we duplicate those images here?

Reviewer 2: Let's do this in a separate PR if you'd like to see this.

Reviewer: #12433

> Instructors may want to group their assessments by course modules (topics, sections or chapters in a course). When using `"groupAssessmentsBy" : "Module"` in `infoCourseInstance.json`, instructors can assign an assessment to a specific module by setting the `module` property in `infoAssessment.json`, as illustrated in the example above. If the property `module` is omitted, by default the assessment will have `"module" : "Default"`.

Reviewer 2: Nit, I think we generally try to use the Oxford comma?

```suggestion
Instructors may want to group their assessments by course modules (topics, sections, or chapters in a course). When using `"groupAssessmentsBy" : "Module"` in `infoCourseInstance.json`, instructors can assign an assessment to a specific module by setting the `module` property in `infoAssessment.json`, as illustrated in the example above. If the property `module` is omitted, by default the assessment will have `"module" : "Default"`.
```

> For more information about how to create your own assessment modules, see [Course configuration](../course/index.md#assessment-modules).

Reviewer 2: Nit:

```suggestion
For more information about how to create your own assessment modules, see the [course configuration](../course/index.md#assessment-modules) documentation.
```

| Assessment type | `shuffleQuestions` default | Meaning                                                                                                                                                                                                                        |
| --------------- | -------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `Homework`      | `false`                    | Homeworks default to using the same question order as in `infoAssessment.json` for all students.                                                                                                                               |
| `Exam`          | `true`                     | Exams default to randomizing the question order individually for each student. This randomization happens within each zone, but the zones themselves always appear in the same order and there is no cross-zone randomization. |

> - If a `Homework` is set to shuffle the question order, it will use a unique-per-course number for each question, so that all students will still get the same question numbers (like #427), but they will not be in order. This makes it easy for students to discuss questions with course staff; they can say “I don't know how to do #427” and everyone will be seeing the same question #427. The main advantage of randomizing question order on Homeworks is to enable data collection on question difficulty and student behavior that is independent of the order in which questions are listed on the assessment.

Reviewer 2: Nit, no need for this to be a list.

```suggestion
If a `Homework` is set to shuffle the question order, it will use a unique-per-course number for each question, so that all students will still get the same question numbers (like #427), but they will not be in order. This makes it easy for students to discuss questions with course staff; they can say “I don't know how to do #427” and everyone will be seeing the same question #427. The main advantage of randomizing question order on Homeworks is to enable data collection on question difficulty and student behavior that is independent of the order in which questions are listed on the assessment.
```

> If a `Homework` is set to shuffle the question order, it will use a unique-per-course number for each question, so that all students will still get the same question numbers (like #427), but they will not be in order. This makes it easy for students to discuss questions with course staff; they can say “I don't know how to do #427” and everyone will be seeing the same question #427. The main advantage of randomizing question order on Homeworks is to enable data collection on question difficulty and student behavior that is independent of the order in which questions are listed on the assessment.

Reviewer: Introducing the concept of a "unique-per-course question number" first would be helpful.

Reviewer 2: Want to do that in another PR?

Reviewer: #12434

| `title`         | string  | The title of the zone. (Optional; default: none)                                                                                                      |
| `questions`     | array   | The list of slots for questions and question alternatives within the zone. (Optional; default: none)                                                  |
| `numberChoose`  | integer | Number of questions to select for each student from this zone. (Optional; default: select all)                                                        |
| `maxPoints`     | number  | Limit on the number of points that can be earned from this zone. (Optional; default: sum of question max points)                                      |
| `bestQuestions` | integer | Only this many questions in the zone will count towards the total points (highest-point questions will count). (Optional; default: use all questions) |
| `comment`       | string  | Free‑form comment for the zone (Optional; default: none)                                                                                              |

> Zone specification details are in the [format specification for `infoAssessment.json`](../schemas/infoAssessment.md)

Author: Done.

- Each zone appears in the given order in the assessment. Zone titles are optional and are displayed to the student if present.

- Within each zone the question order is randomized based on [the assessment type and the `shuffleQuestions` option](#question-order-randomization).

- If a zone has `maxPoints`, then, of the points that are awarded for answering questions in this zone, at most `maxPoints` will count toward the total points.

- If a zone has `bestQuestions`, then, of the questions in this zone, only `bestQuestions` with the highest number of awarded points will count toward the total points.

## Slots for questions and question alternatives

> Each zone has a list of _slots_ given by the `questions` array. Each slot contains either a single question `id`:

Reviewer 2: We're introducing a new term here. Maybe this is the right call given that "alternative groups" or "zone questions" (which we use internally) aren't super friendly names; I just want to make sure we're thinking this through.

Author: The "slot" terminology is one that I often use when giving talks about PL. But I think here it isn't really adding much, and I don't like using terminology in the docs which isn't reflected in the code. I removed "slots" in cdd94bc86d95f807483661a098282fe48aace1fc in favor of talking about "entries in the questions list".

> Each zone has a list of `questions`, each of which can be a single question:

Reviewer 2: Done in bbb2e70.

```json
{ "id": "hardQV1", "points": 10 }
```

Or a slot can contain a _question alternative list_:

```json
{
  "numberChoose": 1,
  "points": 5,
  "alternatives": [{ "id": "FirstAltQ" }, { "id": "SecondAltQ" }]
}
```

> - An assessment question can be specified by either a single `id` or by a list of alternatives, in which case one or more of these alternatives is chosen at random. Once the question `id` is determined, then a random variant of that question is selected. Question alternatives inherit the points of their parent group, if specified.

Reviewer 2: This is the second single-item list I'm seeing; is that intentional for some reason?

Author: Nope, just a result of updating the PR and splitting up a list into multiple parts or deleting all but one item for some reason. Fixed in de53592c59b594e38c8e036f35a362a26d176f9f

}
```

> | Zone Property   | Type    | Description                                                                                                                                           |

Reviewer: A screenshot highlighting what a zone is would be helpful, or more text about what a zone is would be good.

Reviewer 2: Want to add that in another PR?

Reviewer: #12417

| `comment`       | string  | Free‑form comment for the zone. (Optional; default: none)                                                                                             |

Zone specification details are in the [format specification for `infoAssessment.json`](../schemas/infoAssessment.md)

- Each zone appears in the given order in the assessment. Zone titles are optional and are displayed to the student if present.

- Within each zone the question order is randomized based on [the assessment type and the `shuffleQuestions` option](#question-order-randomization).

- If a zone has `maxPoints`, then, of the points that are awarded for answering questions in this zone, at most `maxPoints` will count toward the total points.

- If a zone has `bestQuestions`, then, of the questions in this zone, only `bestQuestions` with the highest number of awarded points will count toward the total points.

## Question alternatives

Each zone has a list of `questions`, each of which can be a single question:

```json
{ "id": "hardQV1", "points": 10 }
```

> Or an entry can contain a _question alternative list_:

Reviewer: Maybe explain what a `question alternative list` is?

```json
{
  "numberChoose": 1,
  "points": 5,
  "alternatives": [{ "id": "FirstAltQ" }, { "id": "SecondAltQ" }]
}
```

If a question alternative list is specified then some of these questions are first selected at random for each student, and then random variants of those questions are generated. Question alternatives inherit the points of their parent group, if specified.

| Slot property     | Type            | Description                                                                                                                                                                 |
| ----------------- | --------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `points`          | number or array | The number of points for this question. Can be a number (e.g., `10`) or a declining list of points (e.g., `[10, 8, 4]`) for partial credit on Exams. (Required; no default) |

Reviewer 2: Definitely; do you want to open a PR to add that after this lands?

> | `points`          | number or array | The number of points for this question. Can be a number (e.g., `10`) or a declining list of points (e.g., `[10, 8, 4]`) for partial credit on Exams. (Required; no default) |

Reviewer 3: #12419.

- If a zone has `bestQuestions`, then, of the questions in this zone, only `bestQuestions` with the highest number of awarded points will count toward the total points.

## Question alternatives

Each zone has a list of `questions`. An entry in this list can be a single question:

```json
{ "id": "hardQV1", "points": 10 }
```

Or an entry can contain a _question alternative list_:

```json
{
  "numberChoose": 1,
  "points": 5,
  "alternatives": [{ "id": "FirstAltQ" }, { "id": "SecondAltQ" }]
}
```

If a question alternative list is specified, some of these questions are first selected at random for each student, and then random variants of those questions are generated. Question alternatives inherit the points of their parent group, if specified.

| Slot property     | Type            | Description                                                                                                                                                                 |
| ----------------- | --------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `points`          | number or array | The number of points for this question. Can be a number (e.g., `10`) or a declining list of points (e.g., `[10, 8, 4]`) for partial credit on Exams. (Required; no default) |

Reviewer 2: In the interest of getting this merged, I'll let you add that in a separate PR 🙂

After the four submissions above the student has achieved 6.5 points for the question, for an overall score of 65% (6.5/10). They can no longer make submission attempts to the question.

## Assessment instances, question selection, and question order

> PrairieLearn distinguishes between _assessments_ and _assessment instances_. An _assessment_ is determined by the code in an `assessments/` directory, and is something like "Midterm 1". Given an assessment, PrairieLearn needs to select the random list of questions for each student, and it is this selection that is the _assessment instance_ for the student. There is only one copy of each assessment, but every student has their own assessment instance.

Reviewer 2: Nit, not that this really matters, but JSON is really more config.

```suggestion
PrairieLearn distinguishes between _assessments_ and _assessment instances_. An _assessment_ is determined by the configuration in an `assessments/` directory, and is something like "Midterm 1". Given an assessment, PrairieLearn needs to select the random list of questions for each student, and it is this selection that is the _assessment instance_ for the student. There is only one copy of each assessment, but every student has their own assessment instance.
```

After the four submissions above the student has achieved 6.5 points for the question, for an overall score of 65% (6.5/10). They can no longer make submission attempts to the question.

## Assessment instances, question selection, and question order

PrairieLearn distinguishes between _assessments_ and _assessment instances_. An _assessment_ is determined by the configuration in an `assessments/` directory, and is something like "Midterm 1". Given an assessment, PrairieLearn needs to select the random list of questions for each student, and it is this selection that is the _assessment instance_ for the student. There is only one copy of each assessment, but every student has their own assessment instance.

> Choosing the selection of questions for each student proceeds in three steps:

Reviewer 2: Changed in bbb2e70.
